src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,960 @@
+    const node_address start = make_node_address( start_pos, driven_veh.face.dir() );
+    known_nodes.emplace( start, make_start_node( start, driven_veh ) );
+    open_set.push( scored_address{ start, 0 } );
+    int search_count = 0;
+    std::vector<std::pair<node_address, navigation_node>> next_nodes;
+    while( !open_set.empty() ) {
+        const node_address cur_addr = open_set.top().addr;
+        open_set.pop();
+        search_count++;
+        if( data.goal_zone.contains( point{ cur_addr.x, cur_addr.y } ) ) {
+            node_address addr = cur_addr;
+            while( !( addr == start ) ) {
+                const navigation_node &node = known_nodes.at( addr );
+                const node_address &prev = node.prev;
+                ret.emplace_back( navigation_step{
+                    nav_corner_pos + point( prev.x, prev.y ),
+                    addr.facing_dir,
+                    node.target_speed_tps *VMIPH_PER_TPS
+                } );
+                addr = prev;
+            }
+            return ret;
+        }
+        const navigation_node &cur_node = known_nodes[cur_addr];
+        next_nodes.clear();
+        compute_next_nodes( cur_addr, cur_node, speed_tps, next_nodes );
+        for( const auto &next : next_nodes ) {
+            const node_address &next_addr = next.first;
+            const navigation_node &next_node = next.second;
+            auto iter = known_nodes.find( next_addr );
+            if( iter != known_nodes.end() ) {
+                const navigation_node &other_node = iter->second;
+                if( next_node.cost < other_node.cost ) {
+                    const bool same_dir = cur_addr.facing_dir == next_addr.facing_dir;
+                    if( ( !same_dir && next_node.tileray_steps == 1 )
+                        || next_addr.facing_dir % 3 == 0 ) {
+                        iter->second = next_node;
+                    }
+                }
+            } else if( known_nodes.size() < max_search_count ) {
+                known_nodes[next_addr] = next_node;
+                open_set.push( compute_node_score( next_addr, next_node ) );
+            }
+        }
+    }
+    return cata::nullopt;
+}
+
+vehicle::autodrive_controller::autodrive_controller( const vehicle &driven_veh,
+        const Character &driver ) : driven_veh( driven_veh ), driver( driver )
+{
+    data.clear();
+}
+
+cata::optional<navigation_step> vehicle::autodrive_controller::compute_next_step()
+{
+    precompute_data();
+    const tripoint_abs_ms veh_pos = abs_vehicle_pos( driven_veh );
+    const bool had_cached_path = !data.path.empty();
+    while( !data.path.empty() && data.path.back().pos != veh_pos ) {
+        data.path.pop_back();
+    }
+    if( data.path.empty() ) {
+        int max_speed_tps = data.max_speed_tps;
+        // if we're just starting out or we've gone off-course use the lowest speed
+        if( had_cached_path || driven_veh.velocity == 0 ) {
+            max_speed_tps = MIN_SPEED_TPS;
```

Comment:
This should also happen if transitioning between land and water (and possibly vice-versa, but the first is more important - the slowdown on entering water causes problems with automatic speed control). (Max speed autodriving should probably also be limited by the safe zone for the engines.)

Reply:
The engine safe speed in what mode, for an amphibious vehicle?

Reply:
Another way to look at it: Since the player can drive their amphibious tank from land to water at any speed without consequences, why should autodrive behave differently?

Reply:
Without consequences? Burning the engines out is not a consequence?

Reply:
Ah, right. I failed to take into account that the transition would change the vehicle's safe_speed. Unlikely to matter at the moment since the max autodrive speed is 12mph, but this could be a problem even on land when going off-road.

Let me think about this some more. I think the right way to handle it is to check the vehicle's safe_speed at every turn, rather than detect any specific terrain transition. After all, you can hit the water at max speed, but the engine won't burn out unless you try to keep running at that speed afterwards.

Reply:
That would also make the check more flexible (able to deal with any other situations in which the maximum safe speed changes). (I think I've seen some amphibious vehicle designs in which the max safe speed in water was below 12 mph, actually.)

Reply:
I think the latest commit should address this now.